File: lib/api/loadouts.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Supervisor contributors. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
import { invoke } from "@tauri-apps/api/core";
import type { Loadout, LoadoutSummary } from "@/types";

export async function listLoadouts(
  gameId: string,
): Promise<[LoadoutSummary[], string]> {
  return invoke("list_loadouts", { gameId });
}

export async function switchLoadout(gameId: string, loadoutId: string): Promise<Loadout> {
  return invoke("switch_loadout", { gameId, loadoutId });
}

export async function createLoadout(gameId: string, name: string): Promise<Loadout> {
  return invoke("create_loadout", { gameId, name });
}

export async function updateLoadout(gameId: string, loadout: Loadout): Promise<Loadout> {
  return invoke("update_loadout", { gameId, loadout });
}

export async function deleteLoadout(gameId: string, loadoutId: string): Promise<void> {
  return invoke("delete_loadout", { gameId, loadoutId });
}

File: src-tauri/src/commands/loadouts.rs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Supervisor contributors. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
use crate::commands::app_data;
use crate::errors::UserFacingIssue;
use crate::loadouts::{Loadout, LoadoutSummary};

#[tauri::command]
pub fn list_loadouts(
    app: tauri::AppHandle,
    game_id: String,
) -> Result<(Vec<LoadoutSummary>, String), UserFacingIssue> {
    let data = app_data(&app)?;
    crate::loadouts::list_loadouts(&data, &game_id).map_err(|e| e.to_user_issue())
}

#[tauri::command]
pub fn switch_loadout(
    app: tauri::AppHandle,
    game_id: String,
    loadout_id: String,
) -> Result<Loadout, UserFacingIssue> {
    let data = app_data(&app)?;
    crate::loadouts::switch_loadout(&data, &game_id, &loadout_id).map_err(|e| e.to_user_issue())
}

#[tauri::command]
pub fn create_loadout(
    app: tauri::AppHandle,
    game_id: String,
    name: String,
) -> Result<Loadout, UserFacingIssue> {
    let data = app_data(&app)?;
    crate::loadouts::create_loadout(&data, &game_id, &name).map_err(|e| e.to_user_issue())
}

#[tauri::command]
pub fn update_loadout(
    app: tauri::AppHandle,
    game_id: String,
    loadout: Loadout,
) -> Result<Loadout, UserFacingIssue> {
    let data = app_data(&app)?;
    crate::loadouts::update_loadout(&data, &game_id, loadout).map_err(|e| e.to_user_issue())
}

#[tauri::command]
pub fn delete_loadout(
    app: tauri::AppHandle,
    game_id: String,
    loadout_id: String,
) -> Result<(), UserFacingIssue> {
    let data = app_data(&app)?;
    crate::loadouts::delete_loadout(&data, &game_id, &loadout_id).map_err(|e| e.to_user_issue())
}
